tests/cpp/test_persistent_buffer.cpp

```diff
@@ -1456,4 +1457,121 @@ TEST_F(PersistentBufferTest, InnerPersistentNotEnoughSharedMemory) {
+  std::vector<int64_t> input_shape{dim0, hidden_size};
+  std::vector<int64_t> norm_shape{hidden_size};
+  auto input_half = makeContigTensor(2, dtype);
+  auto weight_half = makeContigTensor(1, dtype);
+  auto bias_half = makeContigTensor(1, dtype);
+  fusion.addInput(input_half);
+  fusion.addInput(weight_half);
+  fusion.addInput(bias_half);
+  auto input = castOp(DataType::Float, input_half);
+  auto weight = castOp(DataType::Float, weight_half);
+  auto bias = castOp(DataType::Float, bias_half);
+  auto result = layer_norm(input, norm_shape, weight, bias, eps_ptr);
+  auto result_output = castOp(dtype, result.output);
+  fusion.addOutput(result_output);
+  fusion.addOutput(result.mean);
+  fusion.addOutput(result.invstd);
+
+  auto options =
+      at::TensorOptions().dtype(data_type_to_aten(dtype)).device(at::kCUDA, 0);
+  at::Tensor aten_input = at::randn(input_shape, options);
+  c10::optional<at::Tensor> aten_weight = at::randn({input_shape[1]}, options);
+  c10::optional<at::Tensor> aten_bias = at::randn({input_shape[1]}, options);
+  std::vector<c10::IValue> aten_inputs = {aten_input, aten_weight, aten_bias};
+
+  // try translate Welford in fusion
+  KernelArgumentHolder runtime_inputs =
+      KernelArgumentHolder::createKernelArgumentHolder(aten_inputs);
+  SegmentCandidateFinder::translateWelfordInFusion(&fusion, runtime_inputs);
+  auto fusion_copy = fusion;
+
+  // check persistent buffer size
+  SchedulerRuntimeInfo runtime_info(
+      &fusion, {aten_input, aten_weight, aten_bias});
+  auto persistent_buffer_info = scheduler_utils::persistentBuffers(&fusion);
+  auto persistent_buffer_size =
+      persistentBufferSize(&fusion, runtime_info, persistent_buffer_info);
+  int64_t logic_buffer_size = hidden_size * dataTypeSize(dtype);
+  EXPECT_EQ(
+      persistent_buffer_size.projected_persistent_buffer_size,
+      logic_buffer_size);
+
+  // expect segmentation?
+  bool has_enough_regs_smem = true;
+  if (logic_buffer_size > scheduler_utils::register_file_size) {
+    const auto dev_prop = at::cuda::getCurrentDeviceProperties();
+    if ((int64_t)dev_prop->sharedMemPerMultiprocessor < logic_buffer_size) {
+      has_enough_regs_smem = false;
+    } else {
+      int64_t available_buffer_size = normalization_scheduler_utils::
+          getMaxRegOrSharedMemorySizeForPersistentBuffer(
+              &fusion,
+              runtime_info,
+              scheduler_utils::getReductionTvs(&fusion),
+              persistent_buffer_info,
+              /*can_use_smem_persistent*/ true,
+              /*project_to_inputs*/ true);
+      has_enough_regs_smem = available_buffer_size >= logic_buffer_size;
+    }
+  }
+
+  // check segmentation and smem usage
+  FusionExecutorCache executor_cache(std::move(fusion_ptr));
+  auto cg_outputs = executor_cache.runFusionWithInputs(aten_inputs);
+  auto runtime = executor_cache.getMostRecentKernelRuntime();
+  if (has_enough_regs_smem) {
+    // For dtype float, no op scheduler is also used.
```

Comment:
nitpick, can we make this check explicit?

IIUC, half dtype would have a single segment, using inner_persistent;
meanwhile float dtype would give two `no_op + inner_persistent`

Reply:
revised.